akka-stream/src/main/scala/akka/stream/impl/io/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.channels.FileChannel
+
+import akka.actor.{ ActorLogging, DeadLetterSuppression, Props }
+import akka.io.DirectByteBufferPool
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+
+import scala.annotation.tailrec
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object SynchronousFilePublisher {
+  def props(f: File, completionPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, completionPromise, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, bytesReadPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
```

Comment:
I feel we mixed two questions into one somehow. 
1) if the signalled chunk is of known size, 
2) if we read ahead 
(whoa even 3) if this should be a stage or an actor)

The current impl has:
1) known chunk size – the read buffer, someone somewhere will set this value, it's not magic so people could end up depending on it even if not exposed as parameter – (does not concat multiple reads into
2) reads ahead, trying to lessen the delay from request to getting data. Maybe not needed - your points on when it would kick in and may be surprising are good ones.
3) is an Actor. When I started this impl it was the JDK7 thingies (async) and stayed an actor. I can easily make it a Stage, however: in a stage I do not get the `Request(N)` explicitly (unless I missed something in StatefulStage API), and don't know how many chunks to read+collapse.

I don't see why/how I'd want to avoid known chunk size (1), *unless* we do the collapsing we talked about. If we do not read ahead we could "read Demand(N) chunks" collapse them into 1 signal and send it (depleting demand). Now we need to figure out how much we want to read at a given time - so the "how many chunks" needs to be configurable (it's a `maxChunkSize` again) - otherwise I'm stuck reading "forever".

(I need to pick up my car from the workshop, be right back - around 30minutes :-))